tests/31sync/02sync.pl

```diff
@@ -1,3 +1,222 @@
+}
+
+sub matrix_send_room_text_message_and_wait_for_sync
+{
+   my ( $user, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_send_room_text_message( $user, $room_id, %params );
+      },
+      check => sub {
+         my ( $sync_body, $event_id ) = @_;
+
+         sync_timeline_contains( $sync_body, $room_id, sub {
+            $_[0]->{event_id} eq $event_id
+         });
+      },
+   );
+}
+
+sub matrix_send_room_message_and_wait_for_sync
+{
+   my ( $user, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_send_room_message( $user, $room_id, %params );
+      },
+      check => sub {
+         my ( $sync_body, $event_id ) = @_;
+
+         sync_timeline_contains( $sync_body, $room_id, sub {
+            $_[0]->{event_id} eq $event_id
+         });
+      },
+   );
+}
+
+sub matrix_create_room_and_wait_for_sync
+{
+   my ( $user, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_create_room( $user, %params );
+      },
+      check => sub { exists $_[0]->{rooms}{join}{$_[1]} },
+   );
+}
+
+sub matrix_join_room_and_wait_for_sync
+{
+   my ( $user, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_join_room( $user, $room_id, %params );
+      },
+      check => sub { exists $_[0]->{rooms}{join}{$room_id} },
+   );
+}
+
+sub matrix_leave_room_and_wait_for_sync
+{
+   my ( $user, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_leave_room( $user, $room_id, %params );
+      },
+      check => sub { exists $_[0]->{rooms}{leave}{$room_id} },
+   );
+}
+
+sub matrix_put_room_state_and_wait_for_sync
+{
+   my ( $user, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $user,
+      do => sub {
+         matrix_put_room_state( $user, $room_id, %params );
+      },
+      check => sub {
+         my ( $sync_body, $put_result ) = @_;
+         my $event_id = $put_result->{event_id};
+
+         sync_timeline_contains( $sync_body, $room_id, sub {
+            $_[0]->{event_id} eq $event_id;
+         });
+      },
+   );
+}
+
+sub matrix_invite_user_to_room_and_wait_for_sync
+{
+   my ( $inviter, $invitee, $room_id, %params ) = @_;
+
+   matrix_do_and_wait_for_sync( $inviter,
+      do => sub {
+         matrix_do_and_wait_for_sync( $invitee,
+            do => sub {
+               matrix_invite_user_to_room(
+                  $inviter, $invitee, $room_id, %params
+               );
+            },
+            check => sub { exists $_[0]->{rooms}{invite}{$room_id} },
+         );
+      },
+      check => sub {
+         sync_timeline_contains( $_[0], $room_id, sub {
+            $_[0]->{type} eq "m.room.member"
+               and $_[0]->{state_key} eq $invitee->user_id
+               and $_[0]->{content}{membership} eq "invite"
+         });
+      },
+   );
+}
```

Comment:
Not really, otherwise we lose the guarantee that the first sync happens before the action. This would complicate the check.